### PR TITLE
Fixes missing `nth_value` UDAF expr function 

### LINKED ISSUE #12278: NthValue UDAF removed in v41
### Describe the bug

In v41 the `nth_value` expression function API got removed. This happened when `min` and `max` was converted to UDAF in #11013.  This went undetected because there are no logical roundtrip plan tests for `nth_value`.

### To Reproduce

The function is listed in v40:
https://docs.rs/datafusion/40.0.0/datafusion/functions_aggregate/expr_fn/fn.nth_value.html

The function is missing in v41:
https://docs.rs/datafusion/41.0.0/datafusion/functions_aggregate/expr_fn/fn.nth_value.html

### Expected behavior

The `nth_value` UDAF is listed in v41 docs.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jcsherin**
## Which issue does this PR close?

Closes #12278.

## Rationale for this change

Same as #12278.

## What changes are included in this PR?

The API in v40 is:
```rust
pub fn nth_value(args: Vec<Expr>) -> Expr
``` 

A better API would be this:
```rust
pub fn nth_value(
        expr: Expr,
        n: i64,
        order_by: Option<Vec<SortExpr>>
) -> Expr
```

## Are these changes tested?

Yes. Logical roundtrip tests, both with and without order by expression.

## Are there any user-facing changes?

Yes, this is a breaking change of user-facing API.

### CODE REVIEW

File: datafusion/functions-aggregate/src/nth_value.rs [hunk 1]
```diff
@@ -30,19 +30,33 @@ use datafusion_common::{exec_err, internal_err, not_impl_err, Result, ScalarValu
...
+    if let Some(order_by) = order_by {
+        nth_value_udaf()
+            .call(args)
+            .order_by(order_by)
+            .build()
+            .unwrap()
+    } else {
+        nth_value_udaf().call(args)
+    }
+}
```
Comment (jcsherin) [inline datafusion/functions-aggregate/src/nth_value.rs:59]: API changed here.

File: datafusion/functions-aggregate/src/nth_value.rs [hunk 2]
```diff
@@ -30,19 +30,33 @@ use datafusion_common::{exec_err, internal_err, not_impl_err, Result, ScalarValu
...
-    nth_value,
-    "Returns the nth value in a group of values.",
-    nth_value_udaf
-);
+create_func!(NthValueAgg, nth_value_udaf);
+
+/// Returns the nth value in a group of values.
+pub fn nth_value(
+    expr: datafusion_expr::Expr,
+    n: i64,
```
Comment (jayzhan211) [inline datafusion/functions-aggregate/src/nth_value.rs:46]: alternative type is Into<Literal>

File: datafusion/functions-aggregate/src/nth_value.rs [hunk 3]
```diff
@@ -30,19 +30,33 @@ use datafusion_common::{exec_err, internal_err, not_impl_err, Result, ScalarValu
...
+    order_by: Option<Vec<SortExpr>>,
```
Comment (jayzhan211) [inline datafusion/functions-aggregate/src/nth_value.rs:47]: Do we need Option for no ordering case? Can we represent no ordering case with empty vec?

Comment (jcsherin) [reply datafusion/functions-aggregate/src/nth_value.rs:47]: You are right. The option type is extraneous here. I'll make this change.

**Comment by jcsherin** [comment]
I don't have the permission to apply the `api-change` label to this PR.

File: datafusion/functions-aggregate/src/nth_value.rs (hunk 3 above)
Comment (jcsherin) [reply datafusion/functions-aggregate/src/nth_value.rs]: Done.

**Review by jayzhan211 — APPROVED** [review]
👍

**Comment by jayzhan211** [comment]
Thanks @jcsherin

**Pull request merged**